src/Type/EntityStorage/EntityStorageDynamicReturnTypeExtension.php

```diff
@@ -100,6 +88,35 @@ public function getTypeFromMethodCall(
             return $returnType;
         }
 
-        return $type;
+        if (!$callerType instanceof EntityStorageType) {
+            $resolvedEntityType = $this->entityDataRepository->resolveFromStorage($callerType);
+            if ($resolvedEntityType === null) {
+                return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+            }
+            $type = $resolvedEntityType->getClassType();
+        } else {
+            $type = $this->entityDataRepository->get($callerType->getEntityTypeId())->getClassType();
+        }
+
+        if ($type === null) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+        if (\in_array($methodReflection->getName(), ['load', 'loadUnchanged'], true)) {
+            return TypeCombinator::addNull($type);
+        }
+
+        if (\in_array($methodReflection->getName(), ['loadMultiple', 'loadByProperties'], true)) {
+            if ($callerType instanceof ConfigEntityStorageType) {
+                return new ArrayType(new StringType(), $type);
+            }
+
+            return new ArrayType(new IntegerType(), $type);
+        }
+
+        if ($methodReflection->getName() === 'create') {
+            return $type;
+        }
+
+        return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
```

Comment:
I added this to be more explicit. I know we restrict what method names are handled, but it felt weird just dangling the entity class as the return type.